scrapy_poet/injection.py

```diff
@@ -121,6 +121,10 @@ def is_scrapy_response_required(self, request: Request):
         """
         Check whether the request's response is going to be used.
         """
+        # See: https://github.com/scrapinghub/scrapy-poet/issues/48
+        if request.callback is None and request.url.endswith("robots.txt"):
```

Comment:
> I'd prefer not to try to fix Scrapy components on-by-one here; it's a losing battle, as there might be third-party and project-specific extensions which use the same approach.

I agree. There's also the notion of unable to upgrade to the latest Scrapy version with the fix.

> Do you think it would work?

(1) and (2) are good ideas.

I've implemented it in this new commit: https://github.com/scrapinghub/scrapy-poet/pull/106/commits/0ce6de8b0fc64b3acc9d4726ef5664e98f28491b

I'm also thinking if we should introduce an escape hatch for this, but I soon realized that the escape hatch itself is to avoid using the `parse()` method completely. The user warning should be enough to inform the user about this.

On the other hand, it'd hard to make them work for your example [from this comment](https://github.com/scrapinghub/scrapy-poet/issues/48#issuecomment-1341061227) since the `Response` would now actuallly be downloaded but would result in a 404, completely ignoring the `parse()` callback. I think this could be okay, since if users are experiencing 404s, then they could:

- add 404s to the allowed HTTP Response status codes, or
- specify an explicit `callback` and `errback` to handle them.

What do you think?

----

(3) and (4) are all good to me. 👍

> One more question: what about page objects? Is this issue only about DummyResponse? For example, if parse method needs a certain Page Object to be created, would robots.txt middleware trigger creation of this Page Object? What if the page object doesn't use HttpResponse dependency?

The injector still creates them as it's being called by `InjectionMiddleware.process_response`. Good call!

Right now, it doesn't present any problems, apart from some minor wasted compute time to build the dependencies which aren't used at all. It's difficult to address this right now without (3) since simply relying on `request.callback == None` raises some potential false positives.